### PR TITLE
feat: [cds-49192]: Modified stepGroup schema to add support for templates

### PR DESCRIPTION
**Pull request by utkarshchoubey-harness**
Modified stepGroup schema to add support for templates, [ One of (template/steps) ]
Updated the UTs for the same 

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)

### CODE REVIEW

**Review by utkarshchoubey-harness — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919#-:-NJtp7Dz7VmDr0M_LVfk:bb7f2mf)* status: 6 file(s) to review. 0 of 1 approvals obtained. Waiting on team(s): sto.

**Review by brijeshdhakar — DISMISSED** [review]
> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Comment by tchaurasiya** [comment]
*[`953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java` line 51 at r2](https://reviewable.io/reviews/harness/harness-core/41919#-NK-YeZ71qYAXatJjsGq:-NK-YeZ71qYAXatJjsGr:b-78w5jo) ([raw file](https://github.com/harness/harness-core/blob/d8e61b849c97b1851c86f303ef5a543dcee17cde/953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java#L51)):*
> ```Java
> @JsonIgnoreProperties(ignoreUnknown = true)
> @FieldDefaults(level = AccessLevel.PRIVATE)
> @OneOfSet(fields = {"steps", "template"}, requiredFieldNames = {"template"})
> ```

Why does requiredFieldNames only contain "template" here?



_Code quote (from 315-sto-manager/app/src/test/resources/schema/SecurityStage/all.json):_
```JSON
"StepGroupElementConfig"
```

**Review by tchaurasiya — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Review by vivekkarnatiharness — DISMISSED** [review]
> Reviewed 1 of 1 files at r1, 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Review by utkarshchoubey-harness — COMMENTED** [review]
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: 1. unresolved discussions. Waiting on team(s): sto. 

___
*[`953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java` line 51 at r2](https://reviewable.io/reviews/harness/harness-core/41919#-NK-YeZ71qYAXatJjsGq:-NKBv8NNDbBV0VI2ZszS:b-i1trxu) ([raw file](https://github.com/harness/harness-core/blob/d8e61b849c97b1851c86f303ef5a543dcee17cde/953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java#L51)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

Why does requiredFieldNames only contain "template" here?
</blockquote></details>

Done.
As steps field was already notNull therefore it was not required when adding it was showing twice in schema, verified this with Brijesh too

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: 1. unresolved discussions. Waiting on team(s): sto.

**Review by tchaurasiya — COMMENTED** [review]
> Reviewed 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: 1. unresolved discussions. Waiting on team(s): sto.

**Review by tchaurasiya — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

File: 953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java [hunk 1]
```diff
@@ -46,6 +48,7 @@
 @AllArgsConstructor
 @JsonIgnoreProperties(ignoreUnknown = true)
 @FieldDefaults(level = AccessLevel.PRIVATE)
+@OneOfSet(fields = {"steps", "template"}, requiredFieldNames = {"template"})
```
[the remark this answers is not in the document]
Comment (yogesh-chauhan) [reply 953-yaml-commons/src/main/java/io/harness/plancreator/steps/StepGroupElementConfig.java:51]: Discussed offline

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Review by yogesh-chauhan — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Review by harshjain12 — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1, 4 of 5 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919#-:-NKHONk67ZMIFgk21dtz:bnfp4nl)* status: Waiting on team(s): sto.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41919)* status: Waiting on team(s): sto.

**Pull request merged**